### PR TITLE
Use SetHandler fixing mod_rewrite and DirectoryIndex issues

### PR DESCRIPTION
**Pull request by oxyc**
Adding the PR for travis as well as further review and discussion.

- [x] Both dashboard index.html and php_info.php work.
- [x] Default Drupal installation works
  _Status reports only complaint is that trusted hosts settings is disabled._
- [x] Ensure index.html files are used as `DirectoryIndex` #617
  _Added Zen and http://drupalvm.dev/themes/contrib/zen/STARTERKIT/styleguide/ displays `index.html`_
- [x] Fix: Drupal core redirects fail on fresh installation #1055
  > curl -I http://drupalvm.dev/install.php
HTTP/1.1 301 Moved Permanently
  > 
- [x] Drupal 8 not recognizing clean url support #945
  _Adding `install_site: false` and running the installation through `/install.php` adds the `rewrite=ok` GET parameter._

### CODE REVIEW

File: provisioning/templates/nginx-vhost.conf.j2 [hunk 1]
```diff
@@ -34,7 +34,7 @@ server {
         fastcgi_intercept_errors on;
         fastcgi_read_timeout {{ php_max_execution_time }};
         include fastcgi_params;
-        fastcgi_pass {{ php_fpm_listen }};
+        fastcgi_pass unix:{{ php_fpm_listen }};
```
Comment (oxyc) [inline provisioning/templates/nginx-vhost.conf.j2]: Would be nice if we could support both unix and tcp sockets.

Comment (oxyc) [inline provisioning/templates/nginx-vhost.conf.j2]: We could also just stick with tcp sockets. Or are the performance benefits of unix sockets noticeable?

Comment (geerlingguy) [inline provisioning/templates/nginx-vhost.conf.j2]: If possible, I'd like to stick with TCP sockets for now... they are slightly more debuggable (in my experience), and the performance impact is negligible in most real-world use cases. Sockets are very slightly faster, but we're talking < 1ms difference in most Drupal usage.

File: default.config.yml [hunk 2]
```diff
@@ -98,37 +98,36 @@ drupalvm_cron_jobs: []
 # this variable is 'true'.
 configure_drush_aliases: true
 
+apache_vhosts_extra_parameters: |
+    <FilesMatch \.php$>
+      SetHandler "proxy:fcgi://{{ php_fpm_listen }}"
+    </FilesMatch>
```
Comment (oxyc) [inline default.config.yml]: Something like this would be possible. What do you think? Too confusing?

Comment (geerlingguy) [inline default.config.yml]: I think that might be helpful... It's so verbose the way it's set up right now :/

Comment (geerlingguy) [inline default.config.yml]: If you could just add a comment above the variable explaining this is to help configure the PHP-FPM connection to Apache... something like that would be adequate.

**Comment by oxyc** [comment]
Bedtime, I'll test it more thoroughly tomorrow.

**Comment by oxyc** [comment]
@geerlingguy any other scenarios that should be tested? All those issues are now fixed with this PR.

**Comment by geerlingguy** [comment]
> P.S. We're running out of bug/feature issues 👍 

Haha, nice!

**Comment by geerlingguy** [comment]
@oxyc - Does this fix the index.html index for Nginx as well?

**Comment by oxyc** [comment]
Probably not, I thought that nginx worked. I'll provision an nginx box and test it.

**Comment by oxyc** [comment]
Nginx:
- [ ] `/install.php` redirects to `core/install.php`
  _Does not work. You have to visit core/install.php yourself_
- [ ] `rewrite=ok` parameter wasn't added during the install step, but the installation didn't complain about URL rewriting either?
- [x] no complaints from _Status report_
- [ ] Both dashboard index.html and php_info.php work.
  _Erm, the php file didn't work. It was just downloaded. We've forgotten to add the `is_php` attribute on the nginx vhost. If we add it, `index.html` is not shown though._
- [ ] http://drupalvm.dev/themes/contrib/zen/STARTERKIT/styleguide/ displays index.html
  _Nope_

**Comment by oxyc** [comment]
@geerlingguy should we add https://github.com/drupal/drupal/blob/8.3.x/.htaccess#L126 to nginx as well?

One key difference here is that on Apache, you cannot access random php files outside of `core/`. On nginx you can due to this redirect not existing.

**Comment by geerlingguy** [comment]
@oxyc - I would... just because I have had a few people ask my why "install.php" doesn't work (a lot of old scripts/assumptions still use it there and 'just work' in D8 because of that silly redirect :P

**Comment by oxyc** [comment]
This is how I would translate the apache rewrite to nginx.

Apache:
```apache
  RewriteCond %{REQUEST_URI} ^(.*)?/(install.php) [OR]
  RewriteCond %{REQUEST_URI} ^(.*)?/(rebuild.php)
  RewriteCond %{REQUEST_URI} !core
  RewriteRule ^ %1/core/%2 [L,QSA,R=301]
```

nginx:

```nginx
    # Redirect common PHP files to their new locations.
    location ~ ^((?!.*(?:core)).*)/(install.php|rebuild.php) {
         return 301 $scheme://$host$2/core/$1;
    }

# drupalvm.dev/install.php -> drupalvm.dev/core/install.php
# drupalvm.dev/foo/install.php -> drupalvm.dev/foo/core/install.php
```

The problem is we only need this for the drupal vhost, and we need it before passing the file on to fastcgi. `extra_parameters` doesn't work as it's after the fastcgi rule.

**Comment by oxyc** [comment]
This still doesn't add the security measurements the htaccess does. https://github.com/drupal/drupal/blob/8.3.x/.htaccess#L141:L155

**Comment by geerlingguy** [comment]
@oxyc - Yeah... and I don't think we can solve that problem (Drupal community applying security through regex in htaccess files) easily in this PR :)

So I'm okay if a few edge cases aren't working.

**Comment by geerlingguy** [comment]
Otherwise... this looks awesome, and is a great improvement. For some users on really old Ubuntu 12 configurations or still using CentOS 6, this might be the straw that breaks the camel's back—but they can still use the old ProxyPassMatch method (if they desire) by overriding the new var.

Also, existing configs overriding the Vhosts should still work correctly after this merge, so no reason to do a major version bump. It's something that people can take their time to upgrade to (if they need to).

**Comment by geerlingguy** [comment]
So... how many issues did this fix?

- https://github.com/geerlingguy/drupal-vm/issues/617
- https://github.com/geerlingguy/drupal-vm/issues/876
- https://github.com/geerlingguy/drupal-vm/issues/1055

**Pull request merged**